### PR TITLE
[json-rpc] Append uniqueid to PVR.Details.Channel

### PR DESCRIPTION
**Pull request by enen92**
Dear all,

This pull request solves my own feature request made here: http://forum.kodi.tv/showthread.php?tid=244198

Basically, Kodi internally uses the channel uniqueid (for pvr channels) for loads of things. One of them is to match channels to their logos by storing the images in the logo folder in the format: {channelUniqueId}.png|.jpg|.tbn. I edited the wiki with this information some time ago: http://kodi.wiki/view/Settings/Live_TV

Unfortunately it seems there is no way we can grab this information without querying the DB directly. This forces any addon that act as a logo downloader to name the images with the name of the channels to have any chances of working. That adds complexity due to the special characters in some of the channels (and also differs between OS's).

With this pull request I can finally grab the uniqueid of the channels using the json-rpc api, finish my own addon (http://forum.kodi.tv/showthread.php?tid=233068) by downloading the logos with a proper file name and allow people to use thelogodb.com.

No idea who to ping. Pinging @zag2me to act as my defense attorney :)

### CODE REVIEW

File: xbmc/interfaces/json-rpc/schema/types.json [hunk 1]
```diff
@@ -834,7 +834,8 @@
       "thumbnail": { "type": "string" },
       "lastplayed": { "type": "string" },
       "broadcastnow": { "$ref": "PVR.Details.Broadcast" },
-      "broadcastnext": { "$ref": "PVR.Details.Broadcast" }
+      "broadcastnext": { "$ref": "PVR.Details.Broadcast" },
+      "uniqueid": { "type": "integer" }
```
Comment (Razzeee) [inline xbmc/interfaces/json-rpc/schema/types.json]: I think this should be ```{ "$ref": "Library.Id", "required": true }``` but let's wait for @Montellese

**Comment by Razzeee** [comment]
I looked at this, back when I saw your post. But there was some complication I couldn't get solved. Not sure what it was, maybe I looked to deep into the abyss.

File: xbmc/interfaces/json-rpc/schema/types.json (hunk 1 above)
Comment (enen92) [inline xbmc/interfaces/json-rpc/schema/types.json]: Hum I don't know but you already have there a channel id defined that way:
`"channelid": { "$ref": "Library.Id", "required": true },`

Comment (Razzeee) [inline xbmc/interfaces/json-rpc/schema/types.json]: I know, I think Library.Id is another way of saying, this is a positive integer and required says, it should always have data.

**Comment by zag2me** [comment]
Obviously if this improves TV channel logos then great stuff :)  ping the PVR boys just in case they know @ksooo @Jalle19 @xhaggi

**Comment by MartijnKaijser** [comment]
@ksooo yep bump it

**Comment by enen92** [comment]
@ksooo done, I'll rebase later.
Just waiting to see if the `"uniqueid": { "type": "integer" }`is correct

File: xbmc/interfaces/json-rpc/schema/types.json (hunk 1 above)
Comment (enen92) [inline xbmc/interfaces/json-rpc/schema/types.json]: Just compiled a version with your suggestion @Razzeee and it also worked (Y)

File: xbmc/interfaces/json-rpc/schema/types.json [hunk 2]
```diff
@@ -820,7 +820,7 @@
   "PVR.Fields.Channel": {
     "extends": "Item.Fields.Base",
     "items": { "type": "string",
-      "enum": [ "thumbnail", "channeltype", "hidden", "locked", "channel", "lastplayed", "broadcastnow", "broadcastnext" ]
+      "enum": [ "uniqueid", "thumbnail", "channeltype", "hidden", "locked", "channel", "lastplayed", "broadcastnow", "broadcastnext"]
```
Comment (Tolriq) [inline xbmc/interfaces/json-rpc/schema/types.json]: New fields must be added at the end, as JSON specs says (and Montellese), to assure retro compatibility.

File: xbmc/interfaces/json-rpc/schema/types.json (hunk 1 above)
Comment (Tolriq) [inline xbmc/interfaces/json-rpc/schema/types.json]: Library.Id is for the very specific Libraries Id :) Depending on the need, you can add the required true to integer and value limits See https://github.com/xbmc/xbmc/blob/master/xbmc/interfaces/json-rpc/schema/types.json#L204 for example of that.

**Comment by enen92** [comment]
Done and rebased to a single commit

**Comment by Tolriq** [comment]
Version is Major.Minor.Revision. You need to update Minor not Revision, sorry if I was unclear. Revision is for fix without API change, and Major for breaking change in API.

**Comment by enen92** [comment]
Ups :) Hopefully done, sorry.

**Comment by Tolriq** [comment]
Thanks, since quite related to this, and asked by a lot of users, there's channel Id that is the channel id but from Kodi point of view, there's this unique id (where does it come from ?) but is there the original id from the PVR backend ?

Often users says that ids does not correspond to backend but it seems there's no such backend id available ? But maybe this is more for PVR experts.

**Comment by enen92** [comment]
I guess it has to be a pvr expert :)
My understanding of it is that the uniqueid differs from the channel id since it is a function of the channel name and path - getting always a unique value. For instance if you add a channel to a m3u list with a constant name and a constant playpath the uniqueid will always be the same. That does not happen with the channel id. If you:
-remove a known channel from a list and let kodi scrape the list again
-add a few channels to the list and let kodi scrape it again
-add again the known channel

You will notice the channelid that kodi will give to the "known" channel will be different to the older one. Uniqueid, however, will always be the same. I believe that it is the reason the logos are atributted to the channels using the uniqueid and not the channelid.

**Comment by Razzeee** [comment]
Bump
@Montellese for json-rpc
@ksooo and/or @Jalle19 for PVR

**Comment by FernetMenta** [comment]
I don't recommend to expose those kind of internal information. Clients may rely on it and a change in the backend API may also break Kodi API.
Further Kodi's client channel id is an integer and not sufficient to represent the actual IDs used by the some clients. i.e. VDR uses a string to identify a channel. vnsi submits a hash to pvr which would be pretty much useless for display.

**Comment by Tolriq** [comment]
@FernetMenta well users seems to miss the display of the information, from the feedback I have it's purely a display / search thing and maybe a sort thing.

For example in France channel Canal+ is always 4 whatever way you use (Cable box, normal TV, ...). So I suppose they expect to be able to search for 4 to find it or see it in 4th position as it's in their TV.

And currently remotes can only show the name or Kodi id that can be anything but 4.

Maybe there's already some ways, or maybe it should be an optional field used just for information. I just report some recurrent asking.

**Comment by FernetMenta** [comment]
ok, that's a different story. you are asking for client channel number
https://github.com/xbmc/xbmc/blob/master/xbmc/pvr/channels/PVRChannel.h#L290

that should be doable

**Comment by Tolriq** [comment]
So that already exist and just need exposure to JSON that's a good news.

Thanks for information.

**Comment by enen92** [comment]
IMHO there is no reason for having an option to set the channel logos using the channeluniqueId and having absolutely no way of getting that information without looking at the database itself... So probably the identifier that is used to identify a channel for their logos should be exposed by the json-rpc api (this pr) or the usage of the channeluniqueid (https://github.com/xbmc/xbmc/blob/master/xbmc/pvr/channels/PVRChannelGroup.cpp#L334-L337) in the pvr section should be dropped in favor of other identifier that can be easy retrieved by the user. Specially one that does not rely on the channel name (special characters).

**Comment by zag2me** [comment]
Bump... Is this one going in or not?

The Add-on here requires it so its shame to not get it in.

http://forum.kodi.tv/showthread.php?tid=233068&pid=2223903#pid2223903

**Comment by Jalle19** [comment]
Just needs a rebase

**Comment by enen92** [comment]
Unless something went wrong it should be rebased now
Thanks

**Comment by Jalle19** [comment]
jenkins build and merge

**Pull request merged**